iceoryx2/tests/waitset_tests.rs

```diff
@@ -0,0 +1,388 @@
+        let (receiver_1, sender_1) = create_socket();
+        let (receiver_2, _sender_2) = create_socket();
+        let mut guards = vec![];
+
+        guards.push(sut.attach(&listener_1).unwrap());
+        guards.push(sut.attach(&listener_2).unwrap());
+        guards.push(sut.attach(&receiver_1).unwrap());
+        guards.push(sut.attach(&receiver_2).unwrap());
+
+        notifier_1.notify().unwrap();
+        sender_1.try_send(b"bla").unwrap();
+
+        let mut listener_1_triggered = false;
+        let mut listener_2_triggered = false;
+        let mut receiver_1_triggered = false;
+        let mut receiver_2_triggered = false;
+
+        let wait_event = sut
+            .blocking_wait(|attachment_id| {
+                if attachment_id.originates_from(&listener_1) {
+                    listener_1_triggered = true;
+                } else if attachment_id.originates_from(&listener_2) {
+                    listener_2_triggered = true;
+                } else if attachment_id.originates_from(&receiver_1) {
+                    receiver_1_triggered = true;
+                } else if attachment_id.originates_from(&receiver_2) {
+                    receiver_2_triggered = true;
+                } else {
+                    test_fail!("only attachments shall trigger");
+                }
+            })
+            .unwrap();
+
+        assert_that!(wait_event, eq WaitEvent::Notification);
+
+        assert_that!(listener_1_triggered, eq true);
+        assert_that!(receiver_1_triggered, eq true);
+    }
+
+    #[test]
+    fn try_wait_does_not_block<S: Service>()
+    where
+        <S::Event as Event>::Listener: SynchronousMultiplexing,
+    {
+        let _watchdog = Watchdog::new();
+        let node = NodeBuilder::new().create::<S>().unwrap();
+        let sut = WaitSetBuilder::new().create::<S>().unwrap();
+
+        let (listener, _) = create_event::<S>(&node);
+        let _guard = sut.attach(&listener);
+
+        let wait_event = sut
+            .try_wait(|_| {
+                test_fail!("only attachments shall trigger");
+            })
+            .unwrap();
+
+        assert_that!(wait_event, eq WaitEvent::Tick);
+    }
+
+    #[test]
+    fn timed_wait_blocks_for_at_least_timeout<S: Service>()
+    where
+        <S::Event as Event>::Listener: SynchronousMultiplexing,
+    {
+        let _watchdog = Watchdog::new();
+        let node = NodeBuilder::new().create::<S>().unwrap();
+        let sut = WaitSetBuilder::new().create::<S>().unwrap();
+
+        let (listener, _) = create_event::<S>(&node);
+        let _guard = sut.attach(&listener);
+
+        let start = Instant::now();
+        let wait_event = sut
+            .timed_wait(
+                |_| {
+                    test_fail!("only attachments shall trigger");
+                },
+                TIMEOUT,
+            )
+            .unwrap();
+
+        assert_that!(wait_event, eq WaitEvent::Tick);
```

Comment:
I'm wondering whether it would make sense to differentiate between a tick and a timeout. Semantically they are different. The former is expected to happen at some defined interval while the later is a notification that something did not happen.

Reply:
I agree with your point, but I am unsure how to realize this. How could I distinguish between tick and timeout. A tick is emitted every X seconds whilst a timeout is reached after X seconds. So if the tick time is lower than the timeout time, timeout will never be emitted, if it is greater, then timeout and tick will alternate. Mmmmhhh ... I am not sure what to do here...